### PR TITLE
Feature/initialization comp

### PR DESCRIPTION
**Pull request by MichelWU99**
## Description

This PR adds the `auto_initialize_comp` function to the Model class. This function allows finding priors for baseline (e.g level, trend and acceleration), periodic, exponential, white noise and autoregression components. 

### **Model**
I have added the `auto_initialize_comp` function. The user needs to provide a training set as a DataFrame and a ratio of training set that will be used to find the priors, the rest of the training set will be used as a validation set for the priors found. The function checks the `states_name` to know which components need to be studied, then it computes priors through macro-statistic analysis or grid search. Finally it will go through `scipy.minimize` function to determine the best parameters to be used as priors for the Kalman Filter.

### **SKF**
I have added a function just to enable the use the function from `Model`

### **Examples**
I have updated some examples file to execute the `auto_initialize_comp` instead of `auto_initialize_baseline_states`. I have put the `auto_initialize_baseline_states` in comment. 

### **DataProcess**

Change for time for `all_data` to select from the beginning of training to the end of test set.

### **Autoregression Component**

The structure of the noise matrix has been changed to be compatible with `auto_initialize_comp`

## Checklist

- [ ] I have followed the project's coding conventions and style guidelines.
- [ ] I have updated or added documentation where necessary.
- [ ] I have rebased my branch onto the latest `main` (or relevant) branch.
- [ ] I have verified that no sensitive data, plots, or files are included.
- [ ] I have run all previous examples successfully, with no errors.
- [ ] I have tested my changes locally using the following commands/tests:

```shell
pytest
```

## Notes for Reviewers
We still need to look for a way to use user's priors when they are furnished even if they call `auto_initialize_comp` for other components.

### CODE REVIEW

File: src/canari/model.py [hunk 1]
```diff
@@ -33,6 +33,9 @@
 from canari.common import GMA
 from canari.data_process import DataProcess
 from canari.component import Intervention
+import matplotlib.pyplot as plt
+from scipy.optimize import minimize
+from prophet import Prophet
```
Comment (van-dai-vuong) [inline src/canari/model.py:40]:
```suggestion
import logging
```

File: src/canari/model.py [hunk 2]
```diff
@@ -1209,6 +1212,361 @@ def auto_initialize_baseline_states(self, data: np.ndarray):
...
+                                    train_split=ratio_training,
+                                    validation_split=1-ratio_training,
+                                    output_col=[0],
+                                    standardization=False,)
+        
+        train_data,val_data,_,all_data=data_processor.get_splits()
+        mask_s= ~np.isnan(train_data["y"].flatten())
+
+        if "lstm" in self.states_name or "trend" in self.states_name:
+            m = Prophet(
```
Comment (van-dai-vuong) [inline src/canari/model.py]:
```suggestion
            logging.disable(logging.INFO)

            m = Prophet(
```

File: src/canari/model.py [hunk 3]
```diff
@@ -1209,6 +1212,361 @@ def auto_initialize_baseline_states(self, data: np.ndarray):
...
+                # daily_seasonality=True,
+                # n_changepoints=0,
+            )
+            df=pd.DataFrame({"ds": np.asarray(all_data["time"]).reshape(-1),
+                "y":  np.asarray(all_data["y"]).reshape(-1),})
+            m.fit(df)
+            fcst= m.predict(df[["ds"]])
+            trend=fcst["trend"]
+            season_cols = [c for c in ["yearly", "weekly", "daily", "holidays"] if c in fcst.columns]
+            seasonality = fcst[season_cols].sum(axis=1) if season_cols else 0.0
```
Comment (van-dai-vuong) [inline src/canari/model.py:1256]:
```suggestion
            seasonality = fcst[season_cols].sum(axis=1) if season_cols else 0.0

            logging.disable(logging.NOTSET)
            
```

**Comment by van-dai-vuong** [comment]
@MichelWU99 Thank you for preparing the PR. 
(1) Thanks for following my instruction of adding the `auto_initialize_comp` function to necessary files and commented out the old initialization `auto_initialize_baseline_states`. Could you please do it for files in the `test` folder as well? (`test_model_forecast.py` and `test_SKF_anomaly_detection.py`)
(2) Could you add additional examples and unit tests for testing the initializations of `periodic`, `autoregression` and `exponential` components? Because you have already considered states initialization for these components, but in the examples in (1) above, these initializations have never been tested. If you could add them, it will be of great help later on, when I clean `auto_initialize_comp`. 
(3) I noticed that when running files with the new `auto_initialize_comp` function, it prints out the notification when the `scipy.minimizer` starts and finishes. If you accept the suggested code as I commented, it should hide this notification.
(4) Could you add `scipy` and its version to `requirements.txt`?

**Comment by van-dai-vuong** [comment]
(1), (3) have been addressed; (4) will be address in upcoming PR. (2) is remaining.
